app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,266 @@
+    if (timer_cancel_result == -EINPROGRESS) {
+        // too late to cancel, we'll let the timer handler clear up.
+        tap_dance->timer_cancelled = true;
+    }
+    return timer_cancel_result;
+}
+
+static void clear_tap_dance(struct active_tap_dance *tap_dance) {
+    LOG_DBG("Clearing Tap Dance");
+    tap_dance->position = ZMK_BHV_TAP_DANCE_POSITION_FREE;
+    tap_dance->counter = 1;
+}
+
+static void reset_timer(struct active_tap_dance *tap_dance,
+                        struct zmk_behavior_binding_event event) {
+    LOG_DBG("Resetting Timer");
+    // Start the timer.
+    tap_dance->release_at = event.timestamp + tap_dance->config->tapping_term_ms;
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    LOG_DBG("ms_left equal to: %d", ms_left);
+    if (ms_left > 0) {
+        k_delayed_work_submit(&tap_dance->release_timer, K_MSEC(ms_left));
+    }
+}
+
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    LOG_DBG("Press Tap Dance Behavior");
+    struct zmk_behavior_binding binding =
+        (tap_dance->counter < tap_dance->config->behavior_count)
+            ? tap_dance->config->behaviors[(tap_dance->counter) - 1]
+            : tap_dance->config->behaviors[(tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    behavior_keymap_binding_pressed(&binding, event);
+    return 0;
+}
+
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    LOG_DBG("Release Tap Dance Behavior");
+    struct zmk_behavior_binding binding =
+        (tap_dance->counter < tap_dance->config->behavior_count)
+            ? tap_dance->config->behaviors[(tap_dance->counter) - 1]
+            : tap_dance->config->behaviors[(tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    behavior_keymap_binding_released(&binding, event);
+    clear_tap_dance(tap_dance);
+    return 0;
+}
+
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    LOG_DBG("On Binding Pressed");
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        tap_dance = store_tap_dance(event.position, cfg);
+        reset_timer(tap_dance, event);
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = true;
+    stop_timer(tap_dance);
+    if (++tap_dance->counter >= cfg->behavior_count) {
```

Comment:
please put the increment operator outside of the if statement, so it's clear that the increment is happening before the comparison.